cmd/boost/retrieve_cmd.go

```diff
@@ -0,0 +1,468 @@
+package main
+
+import (
+	"context"
+	crand "crypto/rand"
+	"errors"
+	"fmt"
+	"io/fs"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/filecoin-project/boost/lib/keystore"
+	"github.com/filecoin-project/boost/markets/utils"
+	"github.com/filecoin-project/boost/retrieve"
+
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/chain/wallet"
+	lcli "github.com/filecoin-project/lotus/cli"
+
+	"github.com/filecoin-project/go-address"
+
+	"github.com/ipfs/go-bitswap"
+	bsnet "github.com/ipfs/go-bitswap/network"
+	"github.com/ipfs/go-blockservice"
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-datastore"
+	flatfs "github.com/ipfs/go-ds-flatfs"
+	levelds "github.com/ipfs/go-ds-leveldb"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+	offline "github.com/ipfs/go-ipfs-exchange-offline"
+	files "github.com/ipfs/go-ipfs-files"
+	"github.com/ipfs/go-merkledag"
+	unixfile "github.com/ipfs/go-unixfs/file"
+	"github.com/ipld/go-car"
+	"github.com/ipld/go-ipld-prime"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/ipld/go-ipld-prime/node/basicnode"
+	"github.com/ipld/go-ipld-prime/traversal"
+	"github.com/ipld/go-ipld-prime/traversal/selector"
+	"github.com/ipld/go-ipld-prime/traversal/selector/builder"
+	textselector "github.com/ipld/go-ipld-selector-text-lite"
+	"github.com/libp2p/go-libp2p"
+	dht "github.com/libp2p/go-libp2p-kad-dht"
+	"github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/metrics"
+	"github.com/mitchellh/go-homedir"
+	"github.com/urfave/cli/v2"
+	"golang.org/x/xerrors"
+)
+
+var flagMiners = &cli.StringSliceFlag{
+	Name:    "miners",
+	Aliases: []string{"miner", "m"},
+}
+
+var flagOutput = &cli.StringFlag{
+	Name:    "output",
+	Aliases: []string{"o"},
+}
+
+var flagCar = &cli.BoolFlag{
+	Name: "car",
+}
+
+var flagDmPathSel = &cli.StringFlag{
+	Name:  "datamodel-path-selector",
+	Usage: "a rudimentary (DM-level-only) text-path selector, allowing for sub-selection within a deal",
+}
+
+var retrieveCmd = &cli.Command{
+	Name:        "retrieve",
+	Usage:       "Retrieve a file by CID from a miner",
+	Description: "Retrieve a file by CID from a miner. If desired, multiple miners can be specified as fallbacks in case of a failure (comma-separated, no spaces).",
+	ArgsUsage:   "<cid>",
+	Flags: []cli.Flag{
+		flagMiners,
+		flagOutput,
+		flagDmPathSel,
+		flagCar,
+	},
+	Action: func(cctx *cli.Context) error {
+
+		// Store config dir in metadata
+		ddir, err := homedir.Expand("~/.boost-client")
+		if err != nil {
+			fmt.Println("could not set config dir: ", err)
+		}
```

Comment:
Do we also need to call `node.Setup` from cli/node/node.go like we do in boost deal_cmd.go?

Reply:
I've added Blockstore and Datastore to the Node, although just Retrieve needs them. Let me know if you're okay with this... I thought it is a good idea because we do keep state for them in the repository.